diff --git a/pkgs/development/compilers/llvm/common/default.nix b/pkgs/development/compilers/llvm/common/default.nix
index 707bb91b875d..d4b4fb2b5427 100644
--- a/pkgs/development/compilers/llvm/common/default.nix
+++ b/pkgs/development/compilers/llvm/common/default.nix
@@ -780,7 +780,9 @@ let
               "-rtlib=compiler-rt"
               "-Wno-unused-command-line-argument"
               "-B${targetLlvmLibraries.compiler-rt}/lib"
+              "-stdlib=libc++"
             ]
+            ++ lib.optional stdenv.targetPlatform.isLinux "-fuse-ld=mold"
             ++ lib.optional (
               !stdenv.targetPlatform.isWasm && !stdenv.targetPlatform.isFreeBSD
             ) "--unwindlib=libunwind"
@@ -788,11 +790,16 @@ let
               !stdenv.targetPlatform.isWasm
               && !stdenv.targetPlatform.isFreeBSD
               && stdenv.targetPlatform.useLLVM or false
-            ) "-lunwind"
+            ) ["-lunwind" "-lc++"]
             ++ lib.optional stdenv.targetPlatform.isWasm "-fno-exceptions";
           nixSupport.cc-ldflags = lib.optionals (
             !stdenv.targetPlatform.isWasm && !stdenv.targetPlatform.isFreeBSD
-          ) [ "-L${targetLlvmLibraries.libunwind}/lib" ];
+          ) [
+              "-L${targetLlvmLibraries.libunwind}/lib"
+              "-rpath" "${targetLlvmLibraries.libunwind}/lib"
+              "-L${targetLlvmLibraries.libcxx}/lib"
+              "-rpath" "${targetLlvmLibraries.libcxx}/lib"
+            ];
         }
       );
