diff --git a/pkgs/development/compilers/llvm/common/default.nix b/pkgs/development/compilers/llvm/common/default.nix
index 7e8e65b51c1e..6d0162713fd0 100644
--- a/pkgs/development/compilers/llvm/common/default.nix
+++ b/pkgs/development/compilers/llvm/common/default.nix
@@ -428,19 +428,26 @@ let
               "-rtlib=compiler-rt"
               "-Wno-unused-command-line-argument"
               "-B${targetLlvmLibraries.compiler-rt}/lib"
+             "-stdlib=libc++"
             ]
             ++ lib.optional (
               !stdenv.targetPlatform.isWasm && !stdenv.targetPlatform.isFreeBSD
             ) "--unwindlib=libunwind"
-            ++ lib.optional (
+            ++ lib.optionals (
               !stdenv.targetPlatform.isWasm
               && !stdenv.targetPlatform.isFreeBSD
-              && stdenv.targetPlatform.useLLVM or false
-            ) "-lunwind"
+            ) ["-lunwind" "-lc++"]
+            ++ lib.optional (!stdenv.targetPlatform.isDarwin) "-fuse-ld=mold"
             ++ lib.optional stdenv.targetPlatform.isWasm "-fno-exceptions";
           nixSupport.cc-ldflags = lib.optionals (
             !stdenv.targetPlatform.isWasm && !stdenv.targetPlatform.isFreeBSD
-          ) [ "-L${targetLlvmLibraries.libunwind}/lib" ];
+          )
+          [
+            "-L${targetLlvmLibraries.libunwind}/lib"
+            "-rpath" "${targetLlvmLibraries.libunwind}/lib"
+            "-L${targetLlvmLibraries.libcxx}/lib"
+            "-rpath" "${targetLlvmLibraries.libcxx}/lib"
+          ];
         }
       );
