diff --git a/pkgs/tools/admin/trivy/default.nix b/pkgs/tools/admin/trivy/default.nix
index c8cbf6bf8ce2..68e7a7adcc08 100644
--- a/pkgs/tools/admin/trivy/default.nix
+++ b/pkgs/tools/admin/trivy/default.nix
@@ -11,19 +11,23 @@

 buildGoModule rec {
   pname = "trivy";
-  version = "0.52.1";
+  version = "0.52.2";

   src = fetchFromGitHub {
     owner = "aquasecurity";
     repo = "trivy";
     rev = "refs/tags/v${version}";
-    hash = "sha256-7jM0eTUydehzlY8esSxIvQ11Y8VEF6Uyk7mSC0whmq0=";
+    hash = "sha256-3RUL0sgO2/hcfuihNKr51t0qbXvxs9X7yD/OBGATDdw=";
   };

   # Hash mismatch on across Linux and Darwin
   proxyVendor = true;

-  vendorHash = "sha256-MsHlJ56MT+UqF+Psj+hKKpwPnRTcQ3RvWoHbmI93A5E=";
+  vendorHash = "sha256-VkUyjmiiJsDx7NdU6T20LB3tltOYYtf/RaTTPuliMQU=";
+
+  patches = [
+    ./fix_timestamps.diff
+  ];

   subPackages = [ "cmd/trivy" ];

diff --git a/pkgs/tools/admin/trivy/fix_timestamps.diff b/pkgs/tools/admin/trivy/fix_timestamps.diff
new file mode 100644
index 000000000000..44389192f61d
--- /dev/null
+++ b/pkgs/tools/admin/trivy/fix_timestamps.diff
@@ -0,0 +1,13 @@
+diff --git a/pkg/fanal/image/daemon/image.go b/pkg/fanal/image/daemon/image.go
+index 5d80cb93e..3e9c7ae15 100644
+--- a/pkg/fanal/image/daemon/image.go
++++ b/pkg/fanal/image/daemon/image.go
+@@ -110,7 +110,7 @@ func (img *image) ConfigFile() (*v1.ConfigFile, error) {
+ 		return nil, xerrors.Errorf("unable to get diff IDs: %w", err)
+ 	}
+
+-	created, err := time.Parse(time.RFC3339Nano, img.inspect.Created)
++	created, err := time.Parse(time.RFC3339Nano, "0001-01-01T00:00:00.000000000Z")
+ 	if err != nil {
+ 		return nil, xerrors.Errorf("failed parsing created %s: %w", img.inspect.Created, err)
+ 	}
