diff --git a/rust/platform/triple_mappings.bzl b/rust/platform/triple_mappings.bzl
index b6191056..1261b529 100644
--- a/rust/platform/triple_mappings.bzl
+++ b/rust/platform/triple_mappings.bzl
@@ -66,6 +66,8 @@ SUPPORTED_T2_PLATFORM_TRIPLES = {
     "x86_64-unknown-freebsd": _support(std = True, host_tools = True),
     "x86_64-unknown-fuchsia": _support(std = True, host_tools = False),
     "x86_64-unknown-none": _support(std = True, host_tools = False),
+    "x86_64-unknown-linux-musl": _support(std = True, host_tools = False),
+    "aarch64-unknown-linux-musl": _support(std = True, host_tools = False),
 }

 _T3_PLATFORM_TRIPLES = {
