diff --git a/lib/mk-component-set.nix b/lib/mk-component-set.nix
index 90bd180..a8facde 100644
--- a/lib/mk-component-set.nix
+++ b/lib/mk-component-set.nix
@@ -153,10 +153,10 @@ let
             wrap "$dst" ${path + "/pkgs/build-support/bintools-wrapper/ld-wrapper.sh"} "$unwrapped"
           done
         fi
-      ''
-      + optionalString (stdenv.isLinux && pname == "cargo") ''
-        patchelf --add-needed ${pkgsHostHost.libsecret}/lib/libsecret-1.so.0 $out/bin/cargo
       '';
+      # + optionalString (stdenv.isLinux && pname == "cargo") ''
+      #   patchelf --add-needed ${pkgsHostHost.libsecret}/lib/libsecret-1.so.0 $out/bin/cargo
+      # '';

       env = lib.optionalAttrs (pname == "rustc") ({
         inherit (stdenv.cc.bintools) expandResponseParams shell suffixSalt wrapperName coreutils_bin;
